### ISSUE TITLE
Update version of libudis used and add BMI/BMI2

### ISSUE THREAD

**Issue by chrisps** [yegord/snowman#117]
Latest version should be merged from the libudis github. Unfortunately BMI1/BMI2 are not present in it so they need to be added.

**Comment by yegord** [comment]
Would you like to provide a patch?

**Comment by chrisps** [comment]
Sure thing, most of them are pretty simple and don't affect the flags at all. Only problem is I'm not sure what the best approach is with PDEP and PEXT.

**Comment by yegord** [comment]
If you are asking about how to add support for these instructions into udis86, then I do not know: I am not the author of the library, neither I did much reading of its code.

If you are asking about how to define their semantics in InstructionAnalyzer, my suggestion is that you implement them in C, and then rewrite this implementation in IR code. IR statements and expressions have semantics very similar to the one of C (modulo that IR does not do all these int promotions without being asked).

**Comment by chrisps** [comment]
I do understand how to implement it in udis, but inserting an implementation of PDEP/EXT directly into the IR seems like it would make the code less clear to the user as it would insert a new loop into the code. i would rather there be a way to insert an intrinsic of sorts into the code, similar to how Hex-Rays handles it.

**Comment by yegord** [comment]
You do not need to have a loop. The number of bits is limited by a small constant.

Having implementation in IR (rather than a custom intrinsic) has the advantage that, once you have it, all the rest just works: you do not need to worry about type inference rules or value range analysis.

**Comment by chrisps** [comment]
Thing is, PDEP/PEXT accepts non-constant masks. The mask might not be known

**Comment by yegord** [comment]
But the length of the mask is known at IR generation time.
I mean something along the lines of
```
result = 0
flag = ((mask >> 0) & 1)
result = (result << flag) | ((value >> 0) & flag)
flag = ((mask >> 1) & 1)
result = (result << flag) | ((value >> 1) & flag)
...
```

**Comment by chrisps** [comment]
yeah, but the length of the mask is 32 bits in most cases. doesnt that seem like a bit much?

**Comment by yegord** [comment]
I guess, these instructions are not that frequent anyway, so, I would not care.
